deployment/hasura/migrations/AerieMerlin/23_activity_directive_changelog/up.sql

```diff
@@ -0,0 +1,904 @@
+    and revision = (select min(revision)
+                    from activity_directive_changelog
+                    where activity_directive_id = new.activity_directive_id
+                      and plan_id = new.plan_id);
+  return new;
+end$$;
+
+create trigger delete_min_activity_directive_revision_trigger
+  after insert on activity_directive_changelog
+  for each row
+  when (new.revision > 10)
+  execute function delete_min_activity_directive_revision();
+
+create function hasura_functions.restore_activity_changelog(
+  _plan_id integer,
+  _activity_directive_id integer,
+  _revision integer,
+  hasura_session json
+)
+  returns setof activity_directive
+  volatile
+  language plpgsql as $$
+declare
+  _function_permission metadata.permission;
+begin
+  _function_permission :=
+      metadata.get_function_permissions('restore_activity_changelog', hasura_session);
+  if not _function_permission = 'NO_CHECK' then
+    call metadata.check_general_permissions(
+      'restore_activity_changelog',
+      _function_permission, _plan_id,
+      (hasura_session ->> 'x-hasura-user-id')
+    );
+  end if;
+
+  if not exists(select id from public.plan where id = _plan_id) then
+    raise exception 'Plan % does not exist', _plan_id;
+  end if;
+
+  if not exists(select id from public.activity_directive where (id, plan_id) = (_activity_directive_id, _plan_id)) then
+    raise exception 'Activity Directive % does not exist in Plan %', _activity_directive_id, _plan_id;
+  end if;
+
+  if not exists(select revision
+                from public.activity_directive_changelog
+                where (plan_id, activity_directive_id, revision) =
+                      (_plan_id, _activity_directive_id, _revision))
+  then
+    raise exception 'Changelog Revision % does not exist for Plan % and Activity Directive %', _revision, _plan_id, _activity_directive_id;
+  end if;
+
+  return query
+  update activity_directive as ad
+  set name                       = c.name,
+      source_scheduling_goal_id  = c.source_scheduling_goal_id,
+      start_offset               = c.start_offset,
+      type                       = c.type,
+      arguments                  = c.arguments,
+      last_modified_arguments_at = c.changed_arguments_at,
+      metadata                   = c.metadata,
+      anchor_id                  = c.anchor_id,
+      anchored_to_start          = c.anchored_to_start,
+      last_modified_at           = c.changed_at,
+      last_modified_by           = c.changed_by
+  from activity_directive_changelog as c
+  where ad.id                    = _activity_directive_id
+    and c.activity_directive_id  = _activity_directive_id
+    and ad.plan_id               = _plan_id
+    and c.plan_id                = _plan_id
+    and c.revision               = _revision
+  returning ad.*;
+end
+$$;
+
+-- add new function key for restore_activity_changelog
+alter type metadata.function_permission_key add value 'restore_activity_changelog';
```

Comment:
This key needs to be added to the `user` role. [Here](https://github.com/NASA-AMMOS/aerie/blob/6f046fb4c10b31f4210ba30a4b44eca0b61ea0ac/deployment/hasura/migrations/AerieMerlin/23_user_plan_snapshots/up.sql#L504-L509) is how to do so.